File: src/types/image-converter.ts
export type ImageOutputFormat = "png" | "jpeg" | "webp" | "bmp" | "tiff" | "ico";
export type ImageResampleMode = "quality" | "fast";

export interface ImageSourceInfo {
  sourceFormat: string;
  width: number;
  height: number;
  hasAlpha: boolean;
  fileSize: number;
  previewDataUrl: string;
}

export interface ImageConversionOptions {
  targetFormat: ImageOutputFormat;
  quality: number;
  webpLossless: boolean;
  backgroundColor: string;
  icoSize: 16 | 32 | 48 | 64 | 128 | 256 | 512;
}

export interface ImageCompressionOptions {
  targetFormat: "png" | "jpeg" | "webp";
  quality: number;
  webpLossless: boolean;
  backgroundColor: string;
}

export interface ImageCropOptions {
  cropX: number;
  cropY: number;
  cropWidth: number;
  cropHeight: number;
  targetFormat: "png" | "jpeg" | "webp" | "ico";
  quality: number;
  webpLossless: boolean;
  backgroundColor: string;
  icoSize: 16 | 32 | 48 | 64 | 128 | 256 | 512;
  outputSize: number | null;
}

export interface ImageResizeOptions {
  targetWidth: number;
  targetHeight: number;
  targetFormat: "png" | "jpeg" | "webp";
  quality: number;
  webpLossless: boolean;
  backgroundColor: string;
  resampleMode: ImageResampleMode;
}

export interface ImageConversionPreviewResult {
  dataUrl: string;
  previewFormat: string;
  outputFormat: string;
  sourceFormat: string;
  sourceWidth: number;
  sourceHeight: number;
  outputWidth: number;
  outputHeight: number;
  hasAlpha: boolean;
  sourceBytes: number;
  estimatedBytes: number;
}

export const IMAGE_INPUT_EXTENSIONS = ["png", "jpg", "jpeg", "jfif", "webp", "bmp", "tif", "tiff", "ico", "svg"] as const;

export const IMAGE_OUTPUT_FORMATS: { value: ImageOutputFormat; label: string; description: string }[] = [
  { value: "png", label: "PNG", description: "无损，保留透明" },
  { value: "jpeg", label: "JPEG", description: "通用压缩格式" },
  { value: "webp", label: "WebP", description: "现代压缩格式" },
  { value: "bmp", label: "BMP", description: "位图原始格式" },
  { value: "tiff", label: "TIFF", description: "高保真格式" },
  { value: "ico", label: "ICO", description: "桌面图标格式" },
] as const;

export const IMAGE_COMPRESS_FORMATS: { value: ImageCompressionOptions["targetFormat"]; label: string; description: string }[] = [
  { value: "jpeg", label: "JPEG", description: "适合照片和常规资源压缩" },
  { value: "webp", label: "WebP", description: "更现代，通常体积更小" },
  { value: "png", label: "PNG", description: "保留透明，适合 UI 资源" },
] as const;

export const IMAGE_CROP_FORMATS: { value: ImageCropOptions["targetFormat"]; label: string }[] = [
  { value: "png", label: "PNG" },
  { value: "jpeg", label: "JPEG" },
  { value: "webp", label: "WebP" },
  { value: "ico", label: "ICO" },
] as const;

export const IMAGE_RESIZE_FORMATS: { value: ImageResizeOptions["targetFormat"]; label: string }[] = [
  { value: "png", label: "PNG" },
  { value: "jpeg", label: "JPEG" },
  { value: "webp", label: "WebP" },
] as const;

export const ICO_SIZE_OPTIONS = [16, 32, 48, 64, 128, 256, 512] as const;
export const RESIZE_PRESETS = [
  { label: "64", width: 64, height: 64 },
  { label: "128", width: 128, height: 128 },
  { label: "256", width: 256, height: 256 },
  { label: "512", width: 512, height: 512 },
  { label: "1024", width: 1024, height: 1024 },
  { label: "1920×1080", width: 1920, height: 1080 },
] as const;

export const DEFAULT_IMAGE_CONVERSION_OPTIONS: ImageConversionOptions = {
  targetFormat: "png",
  quality: 92,
  webpLossless: false,
  backgroundColor: "#FFFFFF",
  icoSize: 256,
};

export const DEFAULT_IMAGE_COMPRESSION_OPTIONS: ImageCompressionOptions = {
  targetFormat: "webp",
  quality: 80,
  webpLossless: false,
  backgroundColor: "#FFFFFF",
};

export const DEFAULT_IMAGE_CROP_OPTIONS: ImageCropOptions = {
  cropX: 0,
  cropY: 0,
  cropWidth: 256,
  cropHeight: 256,
  targetFormat: "png",
  quality: 92,
  webpLossless: false,
  backgroundColor: "#FFFFFF",
  icoSize: 256,
  outputSize: null,
};

export const DEFAULT_IMAGE_RESIZE_OPTIONS: ImageResizeOptions = {
  targetWidth: 512,
  targetHeight: 512,
  targetFormat: "png",
  quality: 92,
  webpLossless: false,
  backgroundColor: "#FFFFFF",
  resampleMode: "quality",
};

File: src-tauri/src/models/image_converter.rs
//! 图片转换相关数据模型
//!
//! 定义图片转换工具使用的数据结构

use serde::{Deserialize, Serialize};

/// 目标输出格式
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum ImageOutputFormat {
    #[serde(rename = "png")]
    Png,
    #[serde(rename = "jpeg")]
    Jpeg,
    #[serde(rename = "webp")]
    Webp,
    #[serde(rename = "bmp")]
    Bmp,
    #[serde(rename = "tiff")]
    Tiff,
    #[serde(rename = "ico")]
    Ico,
}

impl ImageOutputFormat {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Png => "png",
            Self::Jpeg => "jpeg",
            Self::Webp => "webp",
            Self::Bmp => "bmp",
            Self::Tiff => "tiff",
            Self::Ico => "ico",
        }
    }

    pub fn supports_alpha(&self) -> bool {
        matches!(self, Self::Png | Self::Webp | Self::Tiff | Self::Ico)
    }
}

/// 缩放算法
#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
pub enum ImageResampleMode {
    #[serde(rename = "quality")]
    Quality,
    #[serde(rename = "fast")]
    Fast,
}

/// 图片源信息
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageSourceInfo {
    pub source_format: String,
    pub width: u32,
    pub height: u32,
    pub has_alpha: bool,
    pub file_size: usize,
    pub preview_data_url: String,
}

/// 图片转换配置
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageConversionOptions {
    /// 目标格式
    pub target_format: ImageOutputFormat,

    /// JPEG/WebP 质量，范围 1-100
    #[serde(default = "default_quality")]
    pub quality: u8,

    /// WebP 是否使用无损编码
    #[serde(default)]
    pub webp_lossless: bool,

    /// 透明转不透明时的背景色
    #[serde(default = "default_background_color")]
    pub background_color: String,

    /// ICO 输出尺寸
    #[serde(default = "default_ico_size")]
    pub ico_size: u32,
}

fn default_quality() -> u8 {
    92
}

fn default_background_color() -> String {
    "#FFFFFF".to_string()
}

fn default_ico_size() -> u32 {
    256
}

impl Default for ImageConversionOptions {
    fn default() -> Self {
        Self {
            target_format: ImageOutputFormat::Png,
            quality: default_quality(),
            webp_lossless: false,
            background_color: default_background_color(),
            ico_size: default_ico_size(),
        }
    }
}

/// 图片压缩配置
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageCompressionOptions {
    pub target_format: ImageOutputFormat,
    #[serde(default = "default_quality")]
    pub quality: u8,
    #[serde(default)]
    pub webp_lossless: bool,
    #[serde(default = "default_background_color")]
    pub background_color: String,
}

impl Default for ImageCompressionOptions {
    fn default() -> Self {
        Self {
            target_format: ImageOutputFormat::Webp,
            quality: default_quality(),
            webp_lossless: false,
            background_color: default_background_color(),
        }
    }
}

/// 图片裁剪配置
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageCropOptions {
    pub crop_x: u32,
    pub crop_y: u32,
    pub crop_width: u32,
    pub crop_height: u32,
    pub target_format: ImageOutputFormat,
    #[serde(default = "default_quality")]
    pub quality: u8,
    #[serde(default)]
    pub webp_lossless: bool,
    #[serde(default = "default_background_color")]
    pub background_color: String,
    #[serde(default = "default_ico_size")]
    pub ico_size: u32,
    pub output_size: Option<u32>,
}

/// 图片缩放配置
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageResizeOptions {
    pub target_width: u32,
    pub target_height: u32,
    pub target_format: ImageOutputFormat,
    #[serde(default = "default_quality")]
    pub quality: u8,
    #[serde(default)]
    pub webp_lossless: bool,
    #[serde(default = "default_background_color")]
    pub background_color: String,
    #[serde(default = "default_resample_mode")]
    pub resample_mode: ImageResampleMode,
}

fn default_resample_mode() -> ImageResampleMode {
    ImageResampleMode::Quality
}

/// 图片转换预览结果
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageConversionPreviewResult {
    pub data_url: String,
    pub preview_format: String,
    pub output_format: String,
    pub source_format: String,
    pub source_width: u32,
    pub source_height: u32,
    pub output_width: u32,
    pub output_height: u32,
    pub has_alpha: bool,
    pub source_bytes: usize,
    pub estimated_bytes: usize,
}
